compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp

```diff
@@ -124,22 +197,57 @@ static LogicalResult convertClamp(stablehlo::ClampOp op,
   if (failed(min_or)) return failure();
-  auto operand = op.getOperand();
-  TensorType tensor_type = getCudnnTensorType(operand.getType());
-  auto cast_op = rewriter.create<UnrealizedConversionCastOp>(
-      operand.getLoc(), tensor_type, operand);
+  Type result_type = getCudnnTensorType(op.getType());
+  Value operand = castToCudnnTensor(op.getOperand(), rewriter);
+  Type element_type = op.getType().getElementType();
   rewriter.replaceOpWithNewOp<PointWiseReluOp>(
-      op, tensor_type, cast_op.getResult(0), tensor_type.getElementType(),
+      op, result_type, operand, element_type,
       APFloat(min_or->convertToDouble()));
   return success();
 }
+// Converts convolution 'op' into a cudnn.convolution.
+static LogicalResult convertConv(stablehlo::ConvolutionOp op,
+                                 PatternRewriter& rewriter) {
+  if (!op->getParentOfType<GraphOp>())
+    return rewriter.notifyMatchFailure(op, "expected child of graph");
+  if (failed(matchConv(op, rewriter))) return failure();
+  Type result_type = getCudnnTensorType(op.getType());
+  Value lhs = castToCudnnTensor(op.getLhs(), rewriter);
+  Value rhs = castToCudnnTensor(op.getRhs(), rewriter);
+  Type element_type = op.getType().getElementType();
+
+  APFloat alpha(1.0f);
+  APFloat beta(0.0f);
+
+  uint32_t spatial_dim_count =
+      op.getDimensionNumbers().getInputSpatialDimensions().size();
+  auto get_attr_or = [&](std::optional<DenseIntElementsAttr> attr,
+                         int64_t value) {
+    if (!attr) return SmallVector<int64_t>(spatial_dim_count, value);
+    SmallVector<int64_t> values;
+    llvm::transform(attr->getValues<APInt>(), std::back_inserter(values),
+                    [](APInt stride) { return stride.getSExtValue(); });
+    return values;
+  };
+  SmallVector<int64_t> spatial_stride = get_attr_or(op.getWindowStrides(), 1);
+  SmallVector<int64_t> pre_padding = get_attr_or(op.getPadding(), 0);
+  SmallVector<int64_t> post_padding(spatial_dim_count, 0);
+  SmallVector<int64_t> dilation = get_attr_or(op.getRhsDilation(), 1);
+
+  rewriter.replaceOpWithNewOp<ConvolutionOp>(
+      op, result_type, lhs, rhs, element_type, alpha, beta, spatial_dim_count,
+      spatial_stride, pre_padding, post_padding, dilation);
+  return success();
+}
+
+void populateOutlineHLOToCUDNNPatterns(mlir::RewritePatternSet& patterns) {
+  patterns.add(&outlineClamp);
+  patterns.add(&outlineConv);
+}
 
-void populateHLOToCUDNNPatterns(mlir::RewritePatternSet &patterns) {
-    // TODO(chsigg): don't outline and convert in the same pattern application.
-    // This currently works because the first expects the parent to be a child
-    // of func.func, the second one a child of cudnn.graph.
-    patterns.add(&outlineClamp);
-    patterns.add(&convertClamp);
+void populateConvertHLOToCUDNNPatterns(mlir::RewritePatternSet& patterns) {
+  patterns.add(&convertClamp);
+  patterns.add(&convertConv);
 }
 
 }  // namespace openxla::compiler::nvgpu::cudnn
```

Comment:
missing newline (in vscode you can enable automatic newlines and delete trailing spaces)

Reply:
Thanks for the tip. Done.